server/rest/src/main/java/org/infinispan/rest/resources/ProtobufResource.java

```diff
@@ -0,0 +1,225 @@
+            .create();
+   }
+
+   private CompletionStage<RestResponse> getSchemasNames(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+      return CompletableFuture.supplyAsync(() ->
+          Flowable.fromIterable(cache.keySet())
+               .filter(key -> !((String)key).endsWith(ProtobufMetadataManager.ERRORS_KEY_SUFFIX))
+               .map(key -> {
+                  String error = (String) cache
+                        .get(key + ProtobufMetadataManagerConstants.ERRORS_KEY_SUFFIX);
+                  ProtoSchema protoSchema = new ProtoSchema();
+                  protoSchema.name = (String) key;
+                  if (error != null) {
+                     protoSchema.error = createErrorContent(protoSchema.name, error);
+                  }
+                  return protoSchema;
+               })
+                .sorted(Comparator.comparing(s -> s.name))
+                .collect(Collectors.toList())
+                .map(protoSchemas -> {
+                  try {
+                     byte[] bytes = objectMapper.writeValueAsBytes(protoSchemas);
+                     responseBuilder.contentType(APPLICATION_JSON).entity(bytes);
+                  } catch (JsonProcessingException e) {
+                     throw new RestResponseException(e);
+                  }
+                  return (RestResponse) responseBuilder.build();
+               })
+                .toCompletionStage()
+      , invocationHelper.getExecutor())
+            .thenCompose(Function.identity());
+   }
+
+   private CompletionStage<RestResponse> addOrEditSchema(RestRequest request, boolean create) {
+
+      String schemaName = checkMandatorySchemaName(request);
+
+      ContentSource contents = request.contents();
+      if (contents == null) throw new NoDataFoundException("Schema data not sent in the request");
+
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+      NettyRestResponse.Builder builder = new NettyRestResponse.Builder();
+
+      CompletableFuture<Object> putSchema;
+      if (create) {
+         putSchema = cache.putIfAbsentAsync(schemaName, contents.asString()).thenApply(result -> {
+            if (result == null) {
+               builder.status(HttpResponseStatus.CREATED);
+            } else {
+               builder.status(HttpResponseStatus.CONFLICT);
+            }
+            return result;
+         });
+      } else {
+         putSchema = cache.replaceAsync(schemaName, contents.asString()).thenApply(result -> {
+            if (result == null) {
+               builder.status(HttpResponseStatus.NOT_FOUND);
```

Comment:
This will cause a PUT for a totally new schema to return 404...
PUT should create the schema if it does not exist or replace if it already exists.

Reply:
but if someone has deleted the schema and you are creating it at the same time again ?

Reply:
?

Reply:
for me you use POST to create a total new schema. And then put to modify schemas. this way the action from the user is clear. if you try to modify an already existing schema but the schema does not exist anymore, you get to know

Reply:
nope, PUT should is like AddOrReplace

Reply:
More info: https://tools.ietf.org/html/rfc2616#page-55